### ISSUE TITLE
Default functions

### ISSUE THREAD

**Issue by abelsiqueira** [JuliaSmoothOptimizers/NLPModels.jl#213]
Unless necessary, models should not implement functions that call other functions.
- `grad`, `hprod`, `cons` and similar functions don't need to be implemented because their default is to call their in place version. The same is true for NLS functions
- `op` functions are defined from their `prod` counterpart;
- `jac` and `hess` use `jac_coord` and `hess_coord`;
- `jac_residual` and `hess_residual` should also use `jac_coord_residual` and `hess_coord_residual` (TODO);

**Comment by dpo** [comment]
> jac and hess use jac_coord and hess_coord

Yes, and only `jac_coord!()` and `hess_coord!()` should be implemented. That said, we shouldn't forbid users from implementing, say, `hess()` because some applications may provide the Hessian in CSC format directly, whereas obtaining the sparse coordinate format would require work and computations.

> jac_residual and hess_residual should also use jac_coord_residual and hess_coord_residual

Here too, they should use `jac_coord_residual!()` and `hess_coord_residual!()`.
